File: ui-automation/selenium-java/src/main/java/pages/ShoppingCartPage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import utils.DriverFactory;
import java.util.List;

/**
 * Page Object Model for eCommerce Shopping Cart
 */
public class ShoppingCartPage {
    private WebDriver driver;
    private WebDriverWait wait;
    
    // Page elements
    @FindBy(xpath = "//button[@title='Add to Cart']")
    private WebElement addToCartButton;
    
    @FindBy(xpath = "//table[@class='table']//tbody//tr")
    private List<WebElement> cartItems;
    
    @FindBy(xpath = "//input[@type='number'] | //input[starts-with(@name, 'quantity')]")
    private List<WebElement> quantityInputs;
    
    @FindBy(xpath = "//button[@title='Update']")
    private List<WebElement> updateButtons;
    
    @FindBy(xpath = "//button[@title='Remove']")
    private List<WebElement> removeButtons;
    
    // Constructor
    public ShoppingCartPage() {
        this.driver = DriverFactory.getDriver();
        this.wait = DriverFactory.getWait();
        PageFactory.initElements(driver, this);
    }
    
    // Navigation methods
    public void navigateToProduct(String productId) {
        String url = "https://ecommerce-playground.lambdatest.io/index.php?route=product/product&product_id=" + productId;
        driver.get(url);
    }
    
    public void navigateToCart() {
        driver.get("https://ecommerce-playground.lambdatest.io/index.php?route=checkout/cart");
    }
    
    // Cart operations
    public void addToCart() {
        wait.until(ExpectedConditions.elementToBeClickable(addToCartButton)).click();
    }
    
    public void updateQuantityOfFirstItem(String quantity) {
        WebElement firstRow = cartItems.get(0);
        WebElement quantityInput = firstRow.findElement(
            By.xpath(".//input[@type='number'] | .//input[starts-with(@name, 'quantity')]"));
        WebElement updateButton = firstRow.findElement(By.xpath(".//button[@title='Update']"));
        
        quantityInput.clear();
        quantityInput.sendKeys(quantity);
        updateButton.click();
    }
    
    public void removeSecondItem() {
        if (cartItems.size() >= 2) {
            WebElement secondRow = cartItems.get(1);
            WebElement removeButton = secondRow.findElement(By.xpath(".//button[@title='Remove']"));
            removeButton.click();
        }
    }
    
    // Verification methods
    public int getCartItemCount() {
        return wait.until(ExpectedConditions.visibilityOfAllElements(cartItems)).size();
    }
    
    public String getFirstItemQuantity() {
        WebElement firstRow = cartItems.get(0);
        WebElement quantityInput = firstRow.findElement(
            By.xpath(".//input[@type='number'] | .//input[starts-with(@name, 'quantity')]"));
        return quantityInput.getAttribute("value");
    }
    
    public boolean hasAtLeastTwoItems() {
        return getCartItemCount() >= 2;
    }
}

File: ui-automation/selenium-javascript/pages/ShoppingCartPage.js
const { By, until } = require('selenium-webdriver');

class ShoppingCartPage {
    constructor(driver) {
        this.driver = driver;
        
        // Page element locators
        this.addToCartButton = By.xpath("//button[@title='Add to Cart']");
        this.cartItems = By.xpath("//table[@class='table']//tbody//tr");
        this.quantityInputs = By.xpath("//input[@type='number'] | //input[starts-with(@name, 'quantity')]");
        this.updateButtons = By.xpath("//button[@title='Update']");
        this.removeButtons = By.xpath("//button[@title='Remove']");
    }

    /**
     * Navigate to product page
     */
    async navigateToProduct(productId) {
        const url = `https://ecommerce-playground.lambdatest.io/index.php?route=product/product&product_id=${productId}`;
        await this.driver.get(url);
    }

    /**
     * Navigate to cart page
     */
    async navigateToCart() {
        await this.driver.get('https://ecommerce-playground.lambdatest.io/index.php?route=checkout/cart');
    }

    /**
     * Add product to cart
     */
    async addToCart() {
        const element = await this.driver.wait(until.elementLocated(this.addToCartButton), 10000);
        await this.driver.wait(until.elementIsEnabled(element), 10000);
        await element.click();
    }

    /**
     * Update quantity of first item
     */
    async updateQuantityOfFirstItem(quantity) {
        const cartItemElements = await this.driver.findElements(this.cartItems);
        if (cartItemElements.length > 0) {
            const firstRow = cartItemElements[0];
            
            // Find quantity input in first row
            const quantityInput = await firstRow.findElement(
                By.xpath(".//input[@type='number'] | .//input[starts-with(@name, 'quantity')]")
            );
            
            // Find update button in first row
            const updateButton = await firstRow.findElement(By.xpath(".//button[@title='Update']"));
            
            await quantityInput.clear();
            await quantityInput.sendKeys(quantity);
            await updateButton.click();
        }
    }

    /**
     * Remove second item from cart
     */
    async removeSecondItem() {
        const cartItemElements = await this.driver.findElements(this.cartItems);
        if (cartItemElements.length >= 2) {
            const secondRow = cartItemElements[1];
            const removeButton = await secondRow.findElement(By.xpath(".//button[@title='Remove']"));
            await removeButton.click();
        }
    }

    /**
     * Get cart item count
     */
    async getCartItemCount() {
        try {
            const elements = await this.driver.wait(until.elementsLocated(this.cartItems), 10000);
            return elements.length;
        } catch (error) {
            return 0;
        }
    }

    /**
     * Get first item quantity
     */
    async getFirstItemQuantity() {
        const cartItemElements = await this.driver.findElements(this.cartItems);
        if (cartItemElements.length > 0) {
            const firstRow = cartItemElements[0];
            const quantityInput = await firstRow.findElement(
                By.xpath(".//input[@type='number'] | .//input[starts-with(@name, 'quantity')]")
            );
            return await quantityInput.getAttribute('value');
        }
        return '';
    }

    /**
     * Check if cart has at least two items
     */
    async hasAtLeastTwoItems() {
        return (await this.getCartItemCount()) >= 2;
    }
}

module.exports = ShoppingCartPage;
